runtime/src/execute_cost_table.rs

```diff
@@ -37,64 +48,58 @@ impl ExecuteCostTable {
         }
     }
 
-    pub fn get_cost_table(&self) -> &HashMap<Pubkey, u64> {
-        &self.table
-    }
-
+    // number of programs in table
     pub fn get_count(&self) -> usize {
         self.table.len()
     }
 
-    // instead of assigning unknown program with a configured/hard-coded cost
-    // use average or mode function to make a educated guess.
-    pub fn get_average(&self) -> u64 {
-        if self.table.is_empty() {
-            0
-        } else {
-            self.table.iter().map(|(_, value)| value).sum::<u64>() / self.get_count() as u64
-        }
-    }
-
-    pub fn get_mode(&self) -> u64 {
-        if self.occurrences.is_empty() {
-            0
-        } else {
-            let key = self
-                .occurrences
-                .iter()
-                .max_by_key(|&(_, count)| count)
-                .map(|(key, _)| key)
-                .expect("cannot find mode from cost table");
-
-            *self.table.get(key).unwrap()
-        }
+    // default prorgam cost to max
+    pub fn get_default(&self) -> u64 {
+        // default max comoute units per program
+        200_000u64
     }
 
     // returns None if program doesn't exist in table. In this case,
-    // client is advised to call `get_average()` or `get_mode()` to
-    // assign a 'default' value for new program.
-    pub fn get_cost(&self, key: &Pubkey) -> Option<&u64> {
-        self.table.get(key)
+    // it is advised to call `get_default()` for default program costdefault/
+    // Program cost is estimated as 2 standard deviations above mean, eg
+    // cost = (mean + 2 * std)
+    pub fn get_cost(&self, key: &Pubkey) -> Option<u64> {
+        let aggregated = self.table.get(key)?;
+        Some((aggregated.ema + 2.0 * aggregated.ema_var.sqrt()).ceil() as u64)
```

Comment:
is it possible that `as u64` here would overflow?
would it make sense to change to `u64::try_from` and then use max-cost-units if it fails?

Reply:
> is it possible that `as u64` here would overflow? would it make sense to change to `u64::try_from` and then use max-cost-units if it fails?

good idea!

Reply:
> is it possible that `as u64` here would overflow? would it make sense to change to `u64::try_from` and then use max-cost-units if it fails?

`the trait 'From<f64>' is not implemented for u64`. To prevent u64 overflow, added a simple lossless check when converting ceiling of f64 to u64, should it fail, the function returns None. Test case for it is added.